File: packages/ts/src/signals/hits.ts
/** Anchored signal hit helpers (warehouse shape). */
const SPAN_ANCHORED = new Set([
  "tool_error",
  "llm_error",
  "output_truncated",
  "safety_stop",
  "llm_empty_output",
  "llm_empty_input",
  "io_error_in_output",
  "llm_token_blowup",
]);

export type SignalHit = {
  signal: string;
  kind: string;
  anchor_kind: "span" | "interaction";
  anchor_id: string;
};

export function makeSignalHit(opts: {
  signal: string;
  kind: string;
  anchorKind: "span" | "interaction";
  anchorId: string;
}): SignalHit {
  return {
    signal: opts.signal.trim(),
    kind: opts.kind.trim(),
    anchor_kind: opts.anchorKind,
    anchor_id: opts.anchorId.trim(),
  };
}

export function hitsFromNamedSignals(opts: {
  trigger: Record<string, unknown>[];
  context: Record<string, unknown>[];
  spans: Record<string, unknown>[];
  interactionRef: string;
}): SignalHit[] {
  const { trigger, context, spans, interactionRef } = opts;
  const spansByName = new Map<string, Record<string, unknown>[]>();
  for (const sp of spans) {
    const n = String(sp.name ?? "").trim();
    if (!n) continue;
    const list = spansByName.get(n) ?? [];
    list.push(sp);
    spansByName.set(n, list);
  }
  const pending: SignalHit[] = [];
  const seen = new Set<string>();

  const append = (sig: Record<string, unknown>, kind: string) => {
    const name = String(sig.name ?? "").trim();
    if (!name || seen.has(name)) return;
    seen.add(name);
    let anchorKind: "span" | "interaction" = "interaction";
    let anchorId = interactionRef;
    if (SPAN_ANCHORED.has(name)) {
      anchorKind = "span";
      anchorId = "";
      const toolOrLlm = String(sig.tool ?? sig.llm ?? "").trim();
      if (toolOrLlm) {
        for (const sp of spansByName.get(toolOrLlm) ?? []) {
          anchorId = String(sp.span_ref ?? "");
          if (anchorId) break;
        }
      }
      if (!anchorId && name === "tool_error") {
        for (const sp of spans) {
          if (sp.type === "tool" && sp.status === "failure") {
            anchorId = String(sp.span_ref ?? "");
            break;
          }
        }
      }
      if (
        !anchorId &&
        [
          "llm_error",
          "output_truncated",
          "safety_stop",
          "llm_empty_output",
          "llm_empty_input",
          "io_error_in_output",
          "llm_token_blowup",
        ].includes(name)
      ) {
        for (const sp of spans) {
          if (sp.type !== "llm") continue;
          if (name === "llm_error" && sp.status !== "failure") continue;
          anchorId = String(sp.span_ref ?? "");
          if (anchorId) break;
        }
      }
      if (!anchorId) {
        anchorKind = "interaction";
        anchorId = interactionRef;
      }
    }
    pending.push(makeSignalHit({ signal: name, kind, anchorKind, anchorId }));
  };

  for (const sig of trigger) append(sig, "trigger");
  for (const sig of context) append(sig, "context");
  return pending;
}

File: src/ollie/signals/hits.py
"""Anchored signal hit helpers (warehouse shape)."""

from __future__ import annotations

from typing import Any, Literal

AnchorKind = Literal["span", "interaction"]

_SPAN_ANCHORED = frozenset(
    {
        "tool_error",
        "llm_error",
        "output_truncated",
        "safety_stop",
        "llm_empty_output",
        "llm_empty_input",
        "io_error_in_output",
        "llm_token_blowup",
    }
)


def make_signal_hit(
    *,
    signal: str,
    kind: str,
    anchor_kind: AnchorKind,
    anchor_id: str,
) -> dict[str, Any]:
    return {
        "signal": str(signal).strip(),
        "kind": str(kind).strip(),
        "anchor_kind": anchor_kind,
        "anchor_id": str(anchor_id).strip(),
    }


def hits_from_named_signals(
    *,
    trigger: list[dict[str, Any]],
    context: list[dict[str, Any]],
    spans: list[dict[str, Any]],
    interaction_ref: str,
) -> list[dict[str, Any]]:
    spans_by_name: dict[str, list[dict[str, Any]]] = {}
    for sp in spans:
        if not isinstance(sp, dict):
            continue
        n = str(sp.get("name") or "").strip()
        if n:
            spans_by_name.setdefault(n, []).append(sp)

    pending: list[dict[str, Any]] = []
    seen: set[str] = set()

    def _append(sig: dict[str, Any], kind: str) -> None:
        name = str(sig.get("name") or "").strip()
        if not name or name in seen:
            return
        seen.add(name)
        if name in _SPAN_ANCHORED:
            anchor_kind: AnchorKind = "span"
            anchor_id = ""
            tool_or_llm = str(sig.get("tool") or sig.get("llm") or "").strip()
            if tool_or_llm:
                for sp in spans_by_name.get(tool_or_llm, []):
                    anchor_id = str(sp.get("span_ref") or "")
                    if anchor_id:
                        break
            if not anchor_id and name == "tool_error":
                for sp in spans:
                    if sp.get("type") == "tool" and sp.get("status") == "failure":
                        anchor_id = str(sp.get("span_ref") or "")
                        break
            if not anchor_id and name in ("llm_error", "output_truncated", "safety_stop", "llm_empty_output", "llm_empty_input", "io_error_in_output", "llm_token_blowup"):
                for sp in spans:
                    if sp.get("type") == "llm":
                        if name == "llm_error" and sp.get("status") != "failure":
                            continue
                        anchor_id = str(sp.get("span_ref") or "")
                        if anchor_id:
                            break
            if not anchor_id:
                anchor_kind = "interaction"
                anchor_id = interaction_ref
        else:
            anchor_kind = "interaction"
            anchor_id = interaction_ref
        pending.append(
            make_signal_hit(signal=name, kind=kind, anchor_kind=anchor_kind, anchor_id=anchor_id)
        )

    for sig in trigger:
        if isinstance(sig, dict):
            _append(sig, "trigger")
    for sig in context:
        if isinstance(sig, dict):
            _append(sig, "context")
    return pending
